core/arch/aarch64/clean_call_opt.c

```diff
@@ -88,12 +199,155 @@ analyze_callee_regs_usage(dcontext_t *dcontext, callee_info_t *ci)
      */
 }
 
+/* We use stp/ldp/str/ldr [sp, #imm] pattern to detect callee saved registers,
+ * and assume that the code later won't change those saved value
+ * on the stack.
+ */
 void
 analyze_callee_save_reg(dcontext_t *dcontext, callee_info_t *ci)
 {
-    /* FIXME i#1621: NYI on AArch64
-     * Non-essential for cleancall_opt=1 optimizations.
+    instrlist_t *ilist = ci->ilist;
+    instr_t *top, *bot, *instr;
+    reg_id_t reg1, reg2;
+    bool not_found;
+    /* pointers to instructions of interest */
+    instr_t *enter = NULL, *leave = NULL;
+
+    ci->num_callee_save_regs = 0;
+    top = instrlist_first(ilist);
+    bot = instrlist_last(ilist);
+
+    /* zero or one instruction only, no callee save */
+    if (top == bot)
+        return;
+
+    /* Stack frame analysis
+     * A typical function (fewer than 8 arguments) has the following form:
+     * (a) stp x29, x30, [sp, #-frame_size]!
+     * (b) mov x29, sp
+     * (c) stp x19, x20, [sp, #callee_save_offset]
+     * (c) str x21, [sp, #callee_save_offset+8]
+     * ...
+     * (c) ldp x19, x20, [sp, #callee_save_offset]
+     * (c) ldr x21, [sp, #callee_save_offset+8]
+     * (a) ldp x29, x30, [sp], #frame_size
+     *     ret
+     * Pair (a) appears when the callee calls another function.
+     * If the callee is a leaf function, pair (a) typically has the following form:
+     * (a) sub, sp, sp, #frame_size
+     * (a) add, sp, sp, #frame_size
+     * If (b) is found, x29 is used as the frame pointer.
+     * Pair (c) may have two forms, using stp/ldp for register pairs
+     * or str/ldr for a single callee-saved register.
      */
+     /* Check for pair (a) */
+    for (instr = top; instr != bot; instr = instr_get_next(instr)) {
+        if (instr->bytes == NULL)
+            continue;
+        if (instr_is_push_fp_and_lr(instr) ||
+            instr_is_sub_stk_ptr(instr)) {
+            enter = instr;
+            break;
+        }
+    }
+    if (enter != NULL) {
+        for (instr = bot; instr != enter; instr = instr_get_prev(instr)) {
+            if (!instr->bytes)
+                continue;
+            if (instr_is_pop_fp_and_lr(instr) ||
+                instr_is_add_stk_ptr(instr)) {
+                leave = instr;
+                break;
+            }
+        }
+    }
+    /* Check for (b) */
+    ci->standard_fp = false;
+    if (enter != NULL && leave != NULL &&
+        (ci->bwd_tgt == NULL || instr_get_app_pc(enter) <  ci->bwd_tgt) &&
+        (ci->fwd_tgt == NULL || instr_get_app_pc(leave) >= ci->fwd_tgt)) {
+        for (instr = instr_get_next(enter);
+             instr != leave; instr = instr_get_next(instr)) {
```

Comment:
nit, I usually either put for 3 statements in for in one line or 3 separate lines.